Improve mouse button parsing: helpers and bind{code/sym}

RedSoxFan opened the pull request:
This is the first of five follow-up PRs to #3313. This extracts the parsing of
mouse buttons into helpers and then uses them for `bindcode` and `bindsym`.
The next four PRs are for `input scroll_button`, `seat cursor`, `bar bindsym`,
and `bar tray_bindsym`.

The following helper functions have been added to aid with parsing mouse
buttons from a string:

1. `get_mouse_bindsym`: attempts to parse the string as an x11 button
(button[1-9]) or as an event name (ex BTN_LEFT or BTN_SIDE)
2. `get_mouse_bindcode`: attempts to parse the string as an event code
and validates that the event code is a button (starts with `BTN_`).
3. `get_mouse_button`: this is a conveniency function for callers that
do not care whether a bindsym or bindcode are used and attempts to parse
the string as a bindsym and then bindcode.

This modifies `bindcode` and `bindsym` to use `get_mouse_bindcode` and
`get_mouse_bindsym`, respectively, to parse mouse buttons. Additionally,
the `BINDING_MOUSE` type has been split into `BINDING_MOUSECODE` and
`BINDING_MOUSESYM` to match keys and allow for mouse bindcodes to be
used. Between the two commands, all button syms and codes should be
supported, including x11 axis buttons.

mstoeckl commented:
Just passing by; two comments:
* In this and following PRs, `get_mouse_*` are called with nonnull `error`, in which case the branches testing `if (error)` are unnecessary.
* The code does not account for memory allocation failure, rare as it may be. In general, one should not signal errors using heap allocated values, lest `strdup` or `malloc` return NULL, and the caller falsely assume success. One workaround is a more cumbersome signature, such as `bool parse_x(const char* str, uint32_t* ret, char** err_msg)`.

RedSoxFan commented:
>  In this and following PRs, `get_mouse_*` are called with nonnull `error`, in which case the branches testing `if (error)` are unnecessary.

I originally designed it to make the human readable error message optional. Since all the current callers do request the error message, I have removed the NULL-checks. If the caller doesn't want it, they can just immediately free it.

>  The code does not account for memory allocation failure, rare as it may be. In general, one should not signal errors using heap allocated values, lest `strdup` or `malloc` return NULL, and the caller falsely assume success. One workaround is a more cumbersome signature, such as `bool parse_x(const char* str, uint32_t* ret, char** err_msg)`.

I slightly modified the calling code (for this PR -- I'll update the rest once this is merged) to check for a return value of `0`, which denotes a failure anyway. I also added NULL-checks around the calls to `snprintf` after calls to`malloc`.

ianyfan commented on sway/input/cursor.c:
```diff
@@ -1520,3 +1522,66 @@ void cursor_warp_to_workspace(struct sway_cursor *cursor,
+					"buttons, use the name of the event code.");
+			return 0;
+		}
+		uint32_t buttons[] = {BTN_LEFT, BTN_MIDDLE, BTN_RIGHT,
```
Maybe this could be `static const`?

RedSoxFan replied:
Changed

ianyfan commented:
Thanks!
